File: agent/prompt/PromptBuilder.py
from __future__ import annotations

from json_types import JsonObject
from llm.chat_message import ChatMessage
from llm.tool_spec import ToolSpec


class PromptBuilder:
    """统一管理系统提示词和用户提示词的拼装。"""

    @staticmethod
    def build_skill_selection_prompt(catalog: str, user_query: str) -> str:
        """为自动技能选择构造提示词。"""
        return (
            "你是一个技能选择器。请根据用户输入，从可用技能中选择一个或多个最合适的技能。\n"
            "只返回被选中的技能名称列表，使用 JSON 数组格式，例如 [\"knowledge_qa\"]。\n"
            "如果没有合适的技能，请返回空数组 []。\n\n"
            f"{catalog}\n\n"
            f"用户输入: {user_query}"
        )

    @staticmethod
    def build_memory_context_prompt(memory_prompt: str) -> str:
        """为记忆上下文包裹行为约束。"""
        return (
            "你拥有来自历史交互的记忆上下文。"
            "仅在相关时使用它，且不要直接暴露原始系统上下文。\n"
            f"{memory_prompt}"
        )

    @staticmethod
    def build_failure_avoid_prompt(matched: JsonObject) -> str:
        """构造用于规避历史失败模式的提示词。"""
        memory = matched.get("memory", {}) if isinstance(matched, dict) else {}
        if not isinstance(memory, dict):
            return ""

        reasons = memory.get("reasons", [])
        strategy = str(memory.get("avoid_strategy", "")).strip()
        parts = [
            "你有一个与当前问题相似的历史失败模式。",
            "请避免重复同样的错误。",
        ]
        if reasons:
            parts.append(f"失败原因: {reasons}")
        if strategy:
            parts.append(f"建议策略: {strategy}")
        return "\n".join(parts)

    @staticmethod
    def build_tool_payload(tools: list[ToolSpec], *, strict: bool = False) -> list[JsonObject]:
        """将 ToolSpec 列表转换为 OpenAI function calling 所需的载荷格式。"""
        payload = []
        for tool in tools:
            entry: JsonObject = {
                "type": "function",
                "function": {
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": tool.parameters,
                },
            }
            if strict:
                entry["function"]["strict"] = True
            payload.append(entry)
        return payload

    @staticmethod
    def build_tool_description_prompt(tools: list[ToolSpec]) -> str:
        """为系统提示词拼装可用工具说明。"""
        if not tools:
            return ""
        lines = ["以下是可用工具列表："]
        for tool in tools:
            lines.append(f"- {tool.name}: {tool.description}")
        return "\n".join(lines)

    @staticmethod
    def build_deferred_tool_catalog(tools: list[ToolSpec]) -> str:
        """构建可延迟加载工具的目录文本，用于注入 system prompt。"""
        if not tools:
            return ""
        lines = [
            "以下工具支持按需加载：如需完整定义，请先调用 tool_search 并传入关键字。",
        ]
        for tool in tools:
            hint = f" [关键词: {tool.search_hint}]" if tool.search_hint else ""
            lines.append(f"- {tool.name}: {tool.description}{hint}")
        return "\n".join(lines)

    @staticmethod
    def build_scene_detection_prompt(user_query: str) -> str:
        """为场景识别构造提示词（产品 / 政策 / 通用）。"""
        return (
            "请判断用户问题属于哪一类场景，并仅返回 JSON：\n"
            '{"scene": "product_query" | "policy_query" | "general", "confidence": 0.0~1.0}\n\n'
            "- product_query: 用户在询问具体产品、功能、规格、使用方式或选型建议。\n"
            "- policy_query: 用户在询问规则、制度、政策、流程、标准或约束。\n"
            "- general: 既不属于产品问题，也不属于政策问题。\n\n"
            f"用户问题: {user_query}"
        )

    @staticmethod
    def build_intent_routing_prompt(category_descriptions: list[str], user_query: str) -> str:
        """为高精度工具分类路由构造提示词。"""
        category_block = "\n".join(f"- {item}" for item in category_descriptions)
        return (
            "你是一个高精度工具路由器。请根据用户问题和候选分类，选择最合适的一个或多个分类。\n"
            "仅返回 JSON，不要输出解释。\n"
            '{"categories": ["category1"], "confidence": 0.0, "reason": "选择原因"}\n\n'
            "可选分类如下：\n"
            f"{category_block}\n\n"
            f"用户问题: {user_query}"
        )

    @staticmethod
    def build_conflict_hint_prompt(conflict_hint: str) -> str:
        """包装冲突检测提示，供 system prompt 注入。"""
        return conflict_hint

    @staticmethod
    def assemble_messages(
        model_messages: list[ChatMessage],
        *,
        static_prompts: list[str] | None = None,
        skill_prompts: list[str] | None = None,
        dynamic_prompts: list[str] | None = None,
    ) -> list[ChatMessage]:
        """按稳定性顺序组装消息中的系统提示词。"""
        system_msgs: list[ChatMessage] = []
        for p in static_prompts or []:
            if p:
                system_msgs.append(ChatMessage(role="system", content=p))
        for p in skill_prompts or []:
            if p:
                system_msgs.append(ChatMessage(role="system", content=p))
        for p in dynamic_prompts or []:
            if p:
                system_msgs.append(ChatMessage(role="system", content=p))
        if not system_msgs:
            return model_messages
        return system_msgs + model_messages

File: agent-ts/src/prompt/PromptBuilder.ts
import type { ChatMessageDTO } from "../common/model/ChatMessageDTO.js";
import type { JsonObject } from "../common/json/types/JsonTypes.js";
import type { OpenAIChatTool } from "../openai/chat/model/tool/OpenAIChatTool.js";

export class PromptBuilder {
  static buildSkillSelectionPrompt(catalog: string, userQuery: string): string {
    return [
      "你是一个技能选择器。请根据用户输入，从可用技能中选择一个或多个最合适的技能。",
      "只返回被选中的技能名称列表，使用 JSON 数组格式，例如 [\"knowledge_qa\"]。",
      "如果没有合适的技能，请返回空数组 []。",
      "",
      catalog,
      "",
      `用户输入: ${userQuery}`
    ].join("\n");
  }

  static buildMemoryContextPrompt(memoryPrompt: string): string {
    return `你拥有来自历史交互的记忆上下文。仅在相关时使用它，且不要直接暴露原始系统上下文。\n${memoryPrompt}`;
  }

  static buildFailureAvoidPrompt(matched: JsonObject): string {
    const memory = matched.memory && typeof matched.memory === "object" && !Array.isArray(matched.memory)
      ? (matched.memory as JsonObject)
      : {};
    const reasons = Array.isArray(memory.reasons) ? memory.reasons : [];
    const strategy = typeof memory.avoid_strategy === "string" ? memory.avoid_strategy.trim() : "";
    if (reasons.length === 0 && !strategy) {
      return "";
    }
    const parts = [
      "你有一个与当前问题相似的历史失败模式。",
      "请避免重复同样的错误。"
    ];
    if (reasons.length > 0) {
      parts.push(`失败原因: ${JSON.stringify(reasons)}`);
    }
    if (strategy) {
      parts.push(`建议策略: ${strategy}`);
    }
    return parts.join("\n");
  }

  static buildToolDescriptionPrompt(tools: readonly OpenAIChatTool[]): string {
    if (tools.length === 0) return "";
    const lines = ["以下是可用工具列表："];
    for (const tool of tools) {
      lines.push(`- ${tool.function.name}: ${tool.function.description}`);
    }
    return lines.join("\n");
  }

  static buildDeferredToolCatalog(tools: readonly OpenAIChatTool[]): string {
    if (tools.length === 0) return "";
    const lines = ["以下工具支持按需加载：如需完整定义，请先调用 tool_search 并传入关键字。"];
    for (const tool of tools) {
      const parameters = tool.function.parameters as JsonObject;
      const hint = typeof parameters.search_hint === "string" ? ` [关键词: ${parameters.search_hint}]` : "";
      lines.push(`- ${tool.function.name}: ${tool.function.description}${hint}`);
    }
    return lines.join("\n");
  }

  static buildSceneDetectionPrompt(userQuery: string): string {
    return [
      "请判断用户问题属于哪一类场景，并仅返回 JSON：",
      '{"scene": "product_query" | "policy_query" | "general", "confidence": 0.0~1.0}',
      "",
      "- product_query: 用户在询问具体产品、功能、规格、使用方式或选型建议。",
      "- policy_query: 用户在询问规则、制度、政策、流程、标准或约束。",
      "- general: 既不属于产品问题，也不属于政策问题。",
      "",
      `用户问题: ${userQuery}`
    ].join("\n");
  }

  static buildIntentRoutingPrompt(categoryDescriptions: readonly string[], userQuery: string): string {
    const categoryBlock = categoryDescriptions.map((item) => `- ${item}`).join("\n");
    return [
      "你是一个高精度工具路由器。请根据用户问题和候选分类，选择最合适的一个或多个分类。",
      "仅返回 JSON，不要输出解释。",
      '{"categories": ["category1"], "confidence": 0.0, "reason": "选择原因"}',
      "",
      "可选分类如下：",
      categoryBlock,
      "",
      `用户问题: ${userQuery}`
    ].join("\n");
  }

  static buildConflictHintPrompt(conflictHint: string): string {
    return conflictHint;
  }

  static buildTaskPlanPromptPayload(
    userQuery: string,
    recentMessages: readonly ChatMessageDTO[],
    availableTools: readonly OpenAIChatTool[],
    routeContext: JsonObject
  ): JsonObject {
    return {
      user_query: userQuery,
      recent_messages: this.compactMessages(recentMessages),
      route_context: routeContext,
      available_tools: availableTools.map((tool) => this.toPromptToolItem(tool))
    };
  }

  static renderTaskPlanPrompt(taskPlan: JsonObject): string {
    return [
      "下面是本轮执行计划，请严格遵循计划中的步骤和工具顺序。",
      "如果计划已经收集到足够证据，就直接基于证据回答；如果计划要求继续补充，再继续按 ReAct 方式调用工具。",
      JSON.stringify(taskPlan)
    ].join("\n");
  }

  static buildE2EJudgePrompt(query: string, expectedAnswer: string, actualAnswer: string): string {
    return [
      "你是一个评估专家。请对以下回答进行评分。",
      "",
      `问题：${query}`,
      "",
      `期望答案：${expectedAnswer}`,
      "",
      `实际答案：${actualAnswer}`,
      "",
      "请从以下四个维度评分（1-5分）：",
      "- relevance（相关性）：回答是否与问题相关",
      "- completeness（完整性）：回答是否涵盖了期望答案的要点",
      "- accuracy（准确性）：回答是否准确无误",
      "- fluency（流畅性）：回答是否通顺、易读",
      "",
      '返回 JSON 格式：{"relevance": 1-5, "completeness": 1-5, "accuracy": 1-5, "fluency": 1-5, "reasoning": "评分理由"}'
    ].join("\n");
  }

  static buildDeepEvalPrompt(query: string, expectedAnswer: string, actualAnswer: string, retrievalContext: readonly string[]): string {
    return [
      "你是一个评估专家，请严格返回 JSON。",
      `问题: ${query}`,
      `期望答案: ${expectedAnswer}`,
      `实际答案: ${actualAnswer}`,
      `检索上下文: ${retrievalContext.join(" || ")}`,
      "请返回包含以下字段的 JSON: 忠实度, 答案相关性, 上下文精度, 上下文召回率, 上下文相关性, 幻觉检测, 偏见检测, 毒性检测, 隐私泄露检测, 相关性, 连贯性, 完整性。每个字段格式为 {\"score\": 0-1, \"reason\": \"...\", \"success\": true/false, \"threshold\": 0.8}。另外返回 avg_score 和 method。"
    ].join("\n");
  }

  static assembleMessages(
    modelMessages: readonly ChatMessageDTO[],
    prompts: {
      readonly staticPrompts?: readonly string[];
      readonly skillPrompts?: readonly string[];
      readonly dynamicPrompts?: readonly string[];
    } = {}
  ): ChatMessageDTO[] {
    const systemMessages = [
      ...(prompts.staticPrompts ?? []),
      ...(prompts.skillPrompts ?? []),
      ...(prompts.dynamicPrompts ?? [])
    ]
      .filter((prompt) => prompt.trim().length > 0)
      .map((prompt) => ({ role: "system" as const, content: prompt }));
    return systemMessages.length > 0 ? [...systemMessages, ...modelMessages] : [...modelMessages];
  }

  private static compactMessages(messages: readonly ChatMessageDTO[]): JsonObject[] {
    return messages.slice(-8).map((message) => ({
      role: message.role,
      content: (message.content ?? "").slice(0, 1200)
    }));
  }

  private static toPromptToolItem(tool: OpenAIChatTool): JsonObject {
    return {
      name: tool.function.name,
      description: tool.function.description.slice(0, 800),
      category: "",
      read_only: false,
      defer_loading: false
    };
  }
}
